### PR TITLE
STSMACOM-494 Refactor DateRangeFilter using Final Form

### PR DESCRIPTION
**Pull request by axelhunn**

### CODE REVIEW

**Review by JohnC-80 — COMMENTED** [review]
When I see a third party dependency used in this way, I can't help but have questions about the results... How is behavior with nested RFF `<Form>` components? (I imagine that would happen if someone wanted to use RFF for all of their filters. 
These filter components also serve a searchAndSort alternative, called SearchAndSortQuery (it's in use within the ERM modules and any module that requires more control in the display of their search and results (and wants full control of their own routing))

File: lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js [hunk 1]
```diff
@@ -219,4 +98,4 @@ DateRangeFilter.propTypes = {
   }).isRequired,
 };
 
-export default memo(DateRangeFilter, areSelectedValuesEqual);
+export default DateRangeFilter;
```
[the remark this answers is not in the document]
Comment (axelhunn) [reply lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js:101]: As I understand, the memo with a special comparator was added [here](https://github.com/folio-org/stripes-smart-components/pull/1107/files#diff-7a7fc4e58c62f9302b7679b5ac241c5454b8f984184784ea21822751a88820b6L203-R207) to deal with some kind of infinite state-update-render-effect loop.
This is not needed anymore with final form state management.

File: lib/SearchAndSort/components/DateRangeFilter/Field.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  const { input, meta } = useField(name, { validate });
+
+  return (
+    <Datepicker
+      dateFormat={dateFormat}
+      data-test-date-input={name}
+      usePortal
+
+      // Datepicker doesn't work well with final-form `input` and `meta` API
+      // so we'll use it "ad-hoc" way
```
[the remark this answers is not in the document]
Comment (axelhunn) [reply lib/SearchAndSort/components/DateRangeFilter/Field.js:34]: Sorry, your link points to my code from this PR. Did you mean something different?

[the remark this answers is not in the document]
Comment (axelhunn) [reply lib/SearchAndSort/components/DateRangeFilter/Field.js:34]: I use RFF `useField` hook here which I believe is pretty much the same as using its `<Field>` component.
At least [the doc](https://final-form.org/docs/react-final-form/api/useField) says: "`useField()` is used internally inside `<Field/>`."

File: lib/SearchAndSort/components/DateRangeFilter/Field.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    if (dateIsInvalid) {
+      return (
+        <span data-test-invalid-date={name}>
+          <FormattedMessage id="stripes-smart-components.dateRange.invalidDate" />
+        </span>
+      );
+    }
+
+    return undefined;
```
Comment (vashjs) [inline lib/SearchAndSort/components/DateRangeFilter/Field.js:22]: You can just leave `return;` it's returns `undefined` by default. On line 48 as well

File: lib/SearchAndSort/components/DateRangeFilter/Field.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (axelhunn) [reply lib/SearchAndSort/components/DateRangeFilter/Field.js:34]: Thanks, I will try that

**Review by axelhunn — COMMENTED** [review]
> When I see a third party dependency used in this way, I can't help but have questions about the results... How is behavior with nested RFF `<Form>` components? (I imagine that would happen if someone wanted to use RFF for all of their filters.
> These filter components also serve a searchAndSort alternative, called SearchAndSortQuery (it's in use within the ERM modules and any module that requires more control in the display of their search and results (and wants full control of their own routing))

I think it's safe to guess it's gona be OK with nesting.
The `<Form>` [uses context](https://github.com/final-form/react-final-form/blob/868d8bd73c150e0a5cc68f66270282398baf4274/src/ReactFinalForm.js#L196) for all its doings, and the contexts can be safely nested.

File: lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (axelhunn) [reply lib/SearchAndSort/components/DateRangeFilter/DateRangeFilter.js:101]: Thanks, I will try that

File: lib/SearchAndSort/components/DateRangeFilter/Field.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (axelhunn) [reply lib/SearchAndSort/components/DateRangeFilter/Field.js:34]: Thanks, I will try that

File: lib/SearchAndSort/components/DateRangeFilter/Field.js (hunk 3 above)
Comment (axelhunn) [reply lib/SearchAndSort/components/DateRangeFilter/Field.js:22]: Sonar was not happy with that

**Comment by JohnC-80** [comment]
> I think it's safe to guess it's gona be OK with nesting.
The <Form> uses context for all its doings, and the contexts can be safely nested.

Nesting behavior is all up to how the context itself is used/implemented... Context Providers pass their value props onto child consumers, but there's no automatic aggregation built-in for nested contexts... the value passed to an inner provider will replace the value of an outer provider, unless some aggregation logic is present there.

This component is just a modular piece of the entire filters form, unlike other self-contained forms around the FOLIO system.
`renderFilters` API can allow any declarative wrapping to happen, so long as the SearchAndSort API is met.

If we don't know and we guess, then we'll have to fix our own problem later if it doesn't nest comfortably... this can take a bit of effort, and quite a bit of time if the fix ends up being required in the library we've depended on to fix the original problems for us - [great chance we'll have to implement the solution **for them** and also chance they won't publish the solution until the next breaking version](https://github.com/redux-form/redux-form/pull/4103). (This happened with redux-form on what was a less serious issue, thankfully... but we had console noise for quite some time over that...)

**Comment by axelhunn** [comment]
I don't think we need any aggregation of context. Overriding outer values is just what is needed.
So we use internal Final Form just for internals, and outside the DateRangeFilter exposes only the standard controlled field API like { value, onChange }
This way the outside form state (if any) will be overridden inside. And outside is again exposed only what we choose to expose.

But I agree, we should experiment with that, just to be sure.

**Pull request closed without merging**